lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1002 @@
+	const llvm::DenseMap<mlir::rlc::CA_Value,ConstraintsLattice::IntegerRange>& ConstraintsLattice::getUnderlyingLattice() const{
+		return this->ranges;
+	}
+
+	//TODO: understand what this meet does
+	// In theory it should be the opposite of a join
+	mlir::ChangeResult ConstraintsLattice::meet(const mlir::dataflow::AbstractDenseLattice& val)
+	{
+		//llvm::outs()<<"meet\n";
+		//return mlir::ChangeResult::NoChange;
+		// Do a funny thing and let the meet perform the join
+		return this->join(val);
+	}
+
+	// Simple join operation
+	mlir::ChangeResult ConstraintsLattice::join(const mlir::dataflow::AbstractDenseLattice& val)
+	{
+		//Need to cast it
+		const auto& other = *static_cast<const ConstraintsLattice*>(&val);
+		
+		//Keep it to check if there is no change
+		auto copy = *this;
+
+		// llvm::outs()<<"join\n";
+		// this->print(llvm::outs());
+		// other.print(llvm::outs());
+
+		//Insert the other in this
+		for(const auto& other_range : other.ranges){
+			// We have to insert the other range
+			// If we find it then we have to join, else we have to insert it
+			this->insertOrJoin(other_range.first, other_range.second);
+		}
+
+		//
+		if (*this==copy and other==copy){
+			//this->print(llvm::outs());
+			return mlir::ChangeResult::NoChange;
+		}
+		//this->print(llvm::outs());
+		return mlir::ChangeResult::Change;
+	}
+
+	// Operation to copy two lattices
+	// NB: this is a deep copy
+	mlir::ChangeResult ConstraintsLattice::copy(const ConstraintsLattice& other){
+		
+		// llvm::outs()<<"copy\n";
+		// this->print(llvm::outs());
+		// other.print(llvm::outs());
+
+		if (*this==other){
+			return mlir::ChangeResult::NoChange;
+		}
+
+		//Insert the other in this
+		for(const auto& other_range : other.ranges){
+			// We have to insert the other range
+			// If we find it then we have to join, else we have to insert it
+			this->insertOrUpdate(other_range.first, other_range.second);
+		}
+		
+		return mlir::ChangeResult::Change;
+	}
+
+	// Weird stuff but it works 
+	mlir::ChangeResult ConstraintsLattice::copy(const mlir::dataflow::AbstractDenseLattice& val){
+		return this->copy(*static_cast<const ConstraintsLattice*>(&val));
+	}
+
+	// Printing current lattice (only for debugging)
+	void ConstraintsLattice::print(raw_ostream& os) const{
+
+		os<<"Ranges:\n";
+		if(not this->ranges.empty()){
+			for(auto range: this->ranges){
+				os << "SSA: "<< range.first.value <<" Branch: "<< range.first.return_type << 
+				" RANGE: (" <<range.second.getLower().getSExtValue() << ", " << range.second.getUpper().getSExtValue() << ")\n";
+			}
+		}
+		else {
+			os<<"Empty\n";
+		}
+	}
+
+	bool ConstraintsLattice::operator==(const ConstraintsLattice& other) const{
+		return this->ranges == other.ranges;
+	}
+
+	// Operation called when visiting a region terminator
+	// TODO: maybe remove also this from the lattice
+	bool ConstraintsLattice::setReturnValueInLattice(mlir::RegionBranchOpInterface* branchOp){
```

Comment:
this method should not be a member of constraintLattice, because it depends on yield.